### PR TITLE
ADD test: Able to validate flow with protected path 

### PR DESCRIPTION
**Pull request by andriidovhan**
ADD test: Able to validate flow with protected path

### CODE REVIEW

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudSpec.groovy [hunk 1]
```diff
@@ -437,6 +437,27 @@ class FlowCrudSpec extends BaseSpecification {
         }
     }
 
+    def "Able to validate flow with protected path"() {
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudSpec.groovy]: If this is an all-positive happy path for protected path, I think it worth adding switch validation for each involved switch (in both main and protected paths) as well as traffic examination part. Also try not to overlap with existing tests, I kind of lost count of what is already covered -_-

Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudSpec.groovy]: new step(` when: "Delete rule of protected path on the srcSwitch"`) was added

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudSpec.groovy [hunk 2]
```diff
@@ -437,6 +438,47 @@ class FlowCrudSpec extends BaseSpecification {
...
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flow.allocateProtectedPath = true
+
+        when: "Create a flow with protected path"
+        flowHelper.addFlow(flow)
+
+        then: "Flow with protected path is created"
+        northbound.getFlowPath(flow.id).protectedPath
+
+        and: "Validation of flow with protected path must be succeed"
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudSpec.groovy]: must succeed or must be successful?

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudSpec.groovy (hunk 1 above)
Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudSpec.groovy]: Ok the test is now completely reworked. This is not about FlowCrud anymore. It used to be a positive scenario for creating a flow with protected path but now it is a whole different test that removes rules from the flow. Please see other tests in this spec to comply with the context here. Current implementation is closer to flow validation tests (not flow crud)

Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudSpec.groovy]: sorry for confusing
two separate tests were created
and new validation for each involved switches(main and protected) was added

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudSpec.groovy (hunk 2 above)
Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudSpec.groovy]: fixed

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudSpec.groovy [hunk 3]
```diff
@@ -426,6 +427,79 @@ class FlowCrudSpec extends BaseSpecification {
...
+            allLinks.every { link ->
+                !(link.source.switchId == src.dpId && link.destination.switchId == dst.dpId)
+            } && possibleFlowPaths.size() >= 2 && numberOfDifferentSwitchInPath >= 2
+        } ?: assumeTrue("No suiting switches found", false)
+
+        and: "Create a flow with protected path"
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flow.allocateProtectedPath = true
+
+        when: "Create a flow with protected path"
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudSpec.groovy]: and: "Create a flow with protected path"
  when: "Create a flow with protected path"

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudSpec.groovy [hunk 4]
```diff
@@ -426,6 +427,79 @@ class FlowCrudSpec extends BaseSpecification {
...
+            mainSwitches << topology.activeSwitches.find { it.dpId == flow.source.datapath }
+        }
+        mainSwitches.each { verifySwitchRules(it.dpId) }
+
+        and: "No rule discrepancies on every switch of the flow on the protected path)"
+        def protectedPath = PathHelper.convert(flowPathInfo.protectedPath)
+        def protectedSwitches = pathHelper.getInvolvedSwitches(protectedPath)
+        protectedSwitches.each { verifySwitchRules(it.dpId) }
+
+        when: "Swap path on the flow"
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudSpec.groovy]: You have a separate test doing the same steps, don't you? Or is it different?

Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudSpec.groovy]: I am with you on this one
I will remove this test

Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudSpec.groovy]: I will extend the 'Able to swap flow path' test in the ProtectedPath spec

**Pull request merged**